### DISCUSSION TITLE
DirectWriteForwarder.dll doesn't load on Win7

### DISCUSSION THREAD

**Discussion by Arjunarus** [dotnet/wpf#9158] [Q&A]
Hi, all.
I built wpf 8.0.2 from sources using VS 2022 BuildTools 17.9.2.
Command line is:
dotnet\wpf\8.0.2\build.cmd -configuration Release /p:DotNetUseShippingVersions=true /p:VersionSuffix=<my_suffix> /p:EnableSourceLink=false -verbosity:detailed /m:1

And when i try to use it on Win7 with simple wpf application, I realized that DirectWriteForwarder.dll doesn't want to be load with the error:
"FileLoadException: Could not load file or assembly 'C:\Program Files (x86)\dotnet\shared\Microsoft.WindowsDesktop.App\8.0.2\DirectWriteForwarder.dll'. A dynamic link library (DLL) initialization routine failed. (0x8007045A)"

If i change this DirectWriteForwarder.dll to the original from the official distributiion, then everything is work fine. 
I did some reseearch and discovered that the CLR header of my DirectWriteForwarder.dll and official one are differ.
The offical DirectWriteForwarder.dll has "IL Library" flag in CLR header, but my DirectWriteForwarder.dll has "IL Only" flag. 
Also .text section in the official dll is several times bigger than in mine. The count of sections is also differs: I have 4 sections, but in official are only 3. 
Is somebody know how to build wpf from source to get the similar dll-s like in official distribution?
Or at least how to get DirectWriteForwarder.dll with the same CLR header?
I also extracted that the official DirectWriteForwarder.dll in dotnet 8.0.2 was build fron the same commit 472140dd926227876848e48f41cfc9acb9275492 which I used for my built, this information contained in the official binary. 
Thank you for any help or information.

**Comment by ThomasGoulet73** [comment]
It might be caused by the changes in dotnet/wpf#6538 which made the initialization code of DirectWriteForwarder call a Windows function that might not exist in certain versions of Windows 7. You could try and install [KB2533623](https://support.microsoft.com/kb/2533623) to see if it fixes your issue.

Also note that Windows 7 is no longer supported by Microsoft and thus no longer supported by .Net. See [this link](https://github.com/dotnet/core/blob/main/release-notes/8.0/supported-os.md) for the list of supported OS by .Net 8.

**Reply by Arjunarus** [reply]
Thank you for answer. 
I have Win7 x64 sp1 with all up to dates, but I did not find the KB2533623 in the list of installed KB, also it is not available for download from MS. I found that it was suppressed by the KB3063858, which also is not found in my list of installed KB. So, I downloaded and tried to install KB3063858, but got "the update is not applicable to your computer". I supposed that it was also suppressed by some another KB which is already installed, but I did'nt find the info. 
Also i tried to rollback the #6538 changes in dwriteloader.cpp and rebuilt the DirectWriteForwarder, but the problem is still here. 
I think, if the official and my binary was built from the same commit 472140dd926227876848e48f41cfc9acb9275492 which info I found in the binary, then the official binary also should'n work on my Win7, but it works. 
I still think that the problem is connected to the build process which is somehow differ for the official and my binary. 
If somebody know where could I found the official workflow of building wpf to research it?
May be some env variabes are need to be set or something else?

**Reply by lindexi** [reply]
@Arjunarus Could you try install the KB2533623 ? I give you my download link

- [KB2533623 x86](http://image.acmx.xyz/Windows6.1-KB2533623-x86.msu)
  + MD5： EDF1D538C85F24EC0EF0991E6B27F0D7
  + SHA1： 25BECC0815F3E47B0BA2AE84480E75438C119859
- [KB2533623 x64](http://image.acmx.xyz/Windows6.1-KB2533623-x64.msu)
  + MD5： 0A894C59C245DAD32E6E48ECBDBC8921
  + SHA1： 8A59EA3C7378895791E6CDCA38CC2AD9E83BEBFF
- [KB3063858 32-bit](https://www.microsoft.com/en-us/download/details.aspx?id=47409)
  + MD5： 6d2b63b73e20da5128490632995c4e65
  + SHA1： 3d821d3f47789e11dc3145f436e112762935f126
- [KB3063858 64-bit](https://www.microsoft.com/en-us/download/details.aspx?id=47442)
  + MD5： 6235547a9ac3d931843fe931c15f8e51
  + SHA1： 4ffeb8f40849fda81188e3aad055a30fd9972bd5

 GitHub: [KB2533623.zip](https://github.com/dotnet/wpf/files/15463269/KB2533623.zip)

**Reply by lindexi** [reply]
Why you can not find KB2533623 ? Because [Security update KB2533623 no longer available · Issue #20459 · dotnet/docs](https://github.com/dotnet/docs/issues/20459 )

**Reply by Arjunarus** [reply]
Thank you. I tried to install your KB2533623, but it says "The update is not applicable to your computer". I think maybe it suppressed by some another updates that I already have.

**Reply by lindexi** [reply]
@Arjunarus Could you try KB4457144 ?

**Reply by Arjunarus** [reply]
@lindexi I already tried, it also says "The update is not applicable to your computer".

**Reply by lindexi** [reply]
I can't guess why there's this hint.

**Reply by lindexi** [reply]
And I'm worried there's something else wrong with your system

**Reply by lindexi** [reply]
And the other solution is to rebuild it with https://github.com/Chuyu-Team/YY-Thunks
